### PR TITLE
[#139278341] Fix the refreshToken expiration possibility

### PR DESCRIPTION
**Pull request by paroxp**
## What

As a quick followup to #1, where we've established the refreshToken could expire causing the application to fallout and not log anything, we're implementing more loops to take care of possibility of failing authentication with UAA.

It covers the updating of each running consumer, whenever it's a new client initialisation.

## How to review

- Review the code
- Create yourself an auditor user with UAA
- Either `cf push` or run application locally
  - Before hand, run the following commands [1]
  - Follow the configuration guide in the README
- _Make sure_ to run for ~10 minutes
- _Make sure_ to customise the metric template to test that functionality
- _Make sure_ to create a HostedGraphite trial account and provide the details to the application to test if sending the metrics work - This may take a while to initialise new metrics

[1]: **Configure the UAA expiration dates**

```
$ uaac target "https://api.${DEPLOY_ENV}.dev.cloudpipeline.digital" --skip-ssl-validation
$ uaac token client get admin -s $(aws s3 cp "s3://gds-paas-${DEPLOY_ENV}-state/cf-secrets.yml" -  | awk '/uaa_admin_client_secret/ {print $2}')
$ uaac client update cf --access_token_validity 3 --refresh_token_validity 6
```

## Who can review

Not @combor @keymon or @paroxp

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -130,6 +122,8 @@ func updateApps(client *cfclient.Client, watchedApps map[string]*consumer.Consum
 			}()
 
 			watchedApps[app.Guid] = conn
+		} else if newClient {
+			watchedApps[app.Guid].Stream(app.Guid, authToken)
```
Comment (dcarley) [inline main.go]: Does this cause a new stream to be created? Does it discard the new event and error channels returned from this function? What happens to the existing stream?

Sounds like we need a test for `updateApps(…, true)` :thinking:

File: main_test.go [hunk 2]
```diff
@@ -230,6 +238,83 @@ var _ = Describe("Main", func() {
 			})
 		})
 	})
+
+	Describe("metricProcessor", func() {
+		type tokenJSON struct {
```
Comment (dcarley) [inline main_test.go:241]: Is `oauth.Token` not suitable here?

Comment (keymon) [reply main_test.go:241]: no, it is not. You must use a new type to generate a json that will be parsed properly with a expire date. It must contain `expires` or `expires_in`, and the oauth.Token does not. I will update the comment to explain that.

Comment (paroxp) [reply main_test.go:241]: They're a little different. We've copied this one from the source of [`oauth2.internal`](https://github.com/golang/oauth2/blob/master/internal/token.go#L58-L64).

Comment (keymon) [inline main_test.go]: TBH, checking the code https://github.com/alphagov/paas-cf-apps-statsd/blob/8a7fd4a2e54b0354b0d8ba7bf95c38a6e58e8385/vendor/golang.org/x/oauth2/internal/token.go#L208-L220 maybe if we return a `Content-Type` of `application/x-www-form-urlencoded`, `text/plain` we could use the `oauth.Token`

**Comment by timmow** [comment]
I've tested this and it works ok. I think we should probably squash the WIP commits 😄

**Pull request merged**